feat(binding-coap)!: use new binding-template

JKRhb opened the pull request:
This PR implements a first version of the [CoAP Binding-Template](https://w3c.github.io/wot-binding-templates/bindings/protocols/coap/index.html) in the `binding-coap` package. While all current vocabulary terms are included, not all of them are handled yet (especially the qblockwise and hop-limit parameters).

Due to the changed prefix and a new vocabulary term for the methods, this PR should actually constitute a breaking change. Therefore, it might be included in the 0.9.0 release?

relu91 commented:
> Due to the changed prefix and a new vocabulary term for the methods, this PR should actually constitute a breaking change. Therefore, it might be included in the 0.9.0 release?

> Having said that, we really need a "better" versioning strategy.. w.r.t. to node-wot core but also w.r.t. to the bindings

My understanding is that we are not following semver right now. So we can keep the current rationale that we "manually" decide to switch from 0.x.0 to 0.y.0 . IMHO the right way of fixing this is to set a plan for 1.0.0 and declare that we are in a "beta" stage were we don't follow semver in the README

JKRhb commented on packages/binding-coap/test/coap-client-test.ts:
```diff
@@ -117,7 +117,7 @@ class CoapClientTest {
         const form: CoapForm = {
             href: `coap://127.0.0.1:${port2}`,
-            "cov:methodName": "GET",
+            "coap:method": "GET",
```
Hmm, good question. The [HTTP-in-RDF](https://www.w3.org/TR/HTTP-in-RDF10/) document uses `http` as a prefix. However, the TD document itself [uses `htv`](https://w3c.github.io/wot-thing-description/#http-binding-assertions), so I suppose the default prefix could only be changed in TD 2.0?

danielpeintner replied:
@egekorkan do you know more?

JKRhb replied:
Changed the prefix back to `cov:`

danielpeintner reviewed (commented):
I am not sure if we should approve this PR as is.

@egekorkan mentions in https://github.com/eclipse/thingweb.node-wot/pull/768#discussion_r881801127 he would like to see the `cov` prefix...

egekorkan commented:
Actually I would propose to wait for the editor's draft to become a working draft before going for a major (0.9.0) release since there might be breaking changes in the binding document

danielpeintner commented:
The [CoAP Binding-Template](https://w3c.github.io/wot-binding-templates/bindings/protocols/coap/index.html) document seems to use `cov` prefix again.
Anyhow, it is still not _properly_ published ...

// For the time being I converted the PR to draft.

JKRhb commented:
> The [CoAP Binding-Template](https://w3c.github.io/wot-binding-templates/bindings/protocols/coap/index.html) document seems to use `cov` prefix again. Anyhow, it is still not _properly_ published ...
> 
> // For the time being I converted the PR to draft.

👍

In the meantime, I updated the PR to the draft's latest version and rebased it to the master branch, resolving the conflicts.

egekorkan commented:
FYI: I am planning to get some significant work done on the binding templates and aim for a WD release of the currently well written ones (not hue or xml or json etc.)

ektrah commented on packages/binding-coap/src/coap.ts:
```diff
@@ -29,16 +27,30 @@ export * from "./coap-client";
 }
 
-export type CoapMethodName = "GET" | "POST" | "PUT" | "DELETE" | "FETCH" | "PATCH" | "iPATCH";
+export interface QBlockWiseTransferParameters {
```
`QBlockWiseTransferParameters` and `BlockWiseTransferParameters` have been merged in the latest draft, so there's just `BlockWiseTransferParameters` for both now.

ektrah commented on packages/binding-coap/src/coap.ts:
```diff
@@ -64,3 +76,22 @@ export function isValidCoapMethod(methodName: CoapMethodName): methodName is Coa
+ */
+export function blockSizeToOptionValue(blockSize: BlockSize): BlockSizeOptionValue {
+    // TODO: Check if there is a more elegant solution.
+    const blockSizeOptionValue = Math.floor(Math.sqrt(blockSize) - 4);
```
[Math.log2](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/log2)?

JKRhb replied:
Thanks, adjusted in 501929a0bd76c10d66f3609b35ea70c8689f04d2 :)

JKRhb replied:
You're right, that would actually be unintuitive. I converted the function's implementation in 26f53d026af6ed72e3e6dd60a2ecccbf4fe7fad6 to a switch statement, capturing only the allowed values. Hope that it hasn't become a bit too verbose this way.

ektrah reviewed (commented):
Looks quite good overall; two small comments below.

I think the `form.contentType` to `Content-Format` conversion needs quite a bit more thinking in the spec. Apart from that, I don't expect any major changes at this point.

JKRhb commented on packages/binding-coap/src/coap.ts:
```diff
@@ -29,16 +27,25 @@ export * from "./coap-client";
 export class CoapForm extends Form {
-    public "cov:methodName"?: CoapMethodName;
-    public "cov:options"?: Array<CoapOption> | CoapOption;
+    public "cov:method"?: CoapMethodName;
```
Regarding `cov:method`: As the HTTP binding is using `htv:methodName` (see [here](https://w3c.github.io/wot-binding-templates/bindings/protocols/http/#http-vocabulary-terms)), I was wondering if we should also keep `cov:methodName` for the sake of consistency.

danielpeintner replied:
Consistency is important!
I hope the binding documents do align there as well... which does not seem to be the case.
* https://w3c.github.io/wot-binding-templates/bindings/protocols/http/index.html#http-vocabulary-terms
* https://w3c.github.io/wot-binding-templates/bindings/protocols/coap/index.html#requests-responses

@egekorkan

ektrah replied:
The naming of `cov:method` is consistent with [`modbus:function`](https://w3c.github.io/wot-binding-templates/bindings/protocols/modbus/#form-terms) and [`mqv:controlPacket`](https://w3c.github.io/wot-binding-templates/bindings/protocols/mqtt/#controlpacket), so I don't see any need for change here.

ektrah commented on packages/binding-coap/src/coap-client.ts:
```diff
@@ -231,11 +238,47 @@ export default class CoapClient implements ProtocolClient {
+        const contentFormat = form.contentType;
+        if (contentFormat != null) {
+            debug(`"CoapClient got Form 'contentType' ${contentFormat} `);
+            req.setOption("Content-Format", contentFormat);
```
Let's say there is a property that looks like this:

```json
{
    "op": "readproperty",
    "cov:method": "GET",
    "contentType": "application/json"
}
```

Wouldn't this code set the `Content-Format` option in the GET request to `50`, implying erroneously that the GET request has a payload?

ektrah replied:
Also, since the `contentType` member of a form is mandatory with default value "application/json", I guess the code should look like this:

```diff
-const contentFormat = form.contentType;
-if (contentFormat != null) {
+const contentFormat = form.contentType || "application/json";
 req.setOption("Content-Format", contentFormat);
```

This would mean that *every* request (regardless of method) would have to have a payload, no?

JKRhb replied:
Thank you for the feedback :) I added a couple of updates to the PR. Currently, however, there are a few limitations in node-coap which prevent me from using the numeric values from `cov:contentFormat` and `cov:accept`. I'll try to add these features to node-coap in the upcoming days, then hopefully this PR will be able to be finished :)

JKRhb replied:
I updated node-coap in the meantime, so this thread can be considered resolved :)

JKRhb commented:
> @JKRhb I assume this is now ready for review, right?

Exactly :)

relu91 reviewed (approved):
Thank you @JKRhb! From a quick review, the changes look good. Just a minor comment below and one question:
Regarding [this line](https://github.com/eclipse/thingweb.node-wot/pull/768/files#diff-3f331255e93fb9e3e714530c8787b51ab7615c6cb3ad33a16d1e753d5e5833a7R240) does this means that the coap agent is responsible to translate `contentType` to its CoAP equivalent?

JKRhb commented:
> Thank you @JKRhb! From a quick review, the changes look good. Just a minor comment below and one question: Regarding [this line](https://github.com/eclipse/thingweb.node-wot/pull/768/files#diff-3f331255e93fb9e3e714530c8787b51ab7615c6cb3ad33a16d1e753d5e5833a7R240) does this means that the coap agent is responsible to translate `contentType` to its CoAP equivalent?

Thank you for the feedback :) The current procedure is as follows: If `form["cov:contentFormat"]` is defined, then the client will simply use the numeric value. Otherwise, it will use the string representation in the `contentType` field or `application/json` as a fallback and try to map it to a Content-Format, based on the (currently somewhat flawed) logic implemented in node-coap.

relu91 commented:
Ok! thanks again :) let's merge this.
